Documentation/allow_legacy_mode.md

```diff
@@ -0,0 +1,101 @@
+
+etcd v2.0 uses internal API v2 protocol, which is not compatible with the v1 protocol.
+
+Allow-legacy mode is introduced to help user upgrade v0.4 etcd cluster easily, and ensures cluster has a short unavailable period, which is less than 1 minute in most cases (based on the size of the actual data).
+
+### Setup
+
+This mode is on if `ETCD_ALLOW_LEGACY_MODE` is set to true, or etcd is running in CoreOS system.
+
+It treats `ETCD_BINARY_DIR` as the directory for etcd binaries, which is organized in this way:
+
+```
+ETCD_BINARY_DIR
+  |
+  -- 1
+  |
+  -- 2
+```
+
+`1` is etcd with internal API v1 protocol. You should use etcd v0.4.7 here. `2` is etcd with internal API v2 protocol, which is etcd v2.0.
+
+The default value for `ETCD_BINARY_DIR` is `/usr/libexec/etcd/internal_versions/`.
+
+### Usage
+
+When bootstrapping with v1 data directory and v1 flags, etcd executes v0.4.7 binary and runs exactly the same as before.
+
+Before upgrade, you SHOULD check the health of the cluster by running:
+
+```
+$ etcdctl cluster-health
+```
+
+If the cluster and all members are healthy, you can start the upgrading process. Or you should check the unhealthy machines and fix them.
+
+
+Call upgrade command to upgrade etcd cluster to v2.0 now:
+
+```
+$ etcdctl upgrade --old-version=1 --new-version=2 --peer-url=$PEER_URL
+```
+
+`PEER_URL` is any accessible peer url of the cluster.
+
+All peer-mode members will print out:
+
+```
+detected next internal version 2, exit after 10 seconds.
+```
+
+etcd does some internal coordination within the cluster in this period and then exits.
+
+You need to restart them after they exit. When restart, it upgrades the data directory, and executes etcd v2.0. If it runs in CoreOS system, the init system can take care of restart for you.
+
+After the upgrade process, you can run the health check again to verify the upgrade. If the cluster is unhealthy or there is an unhealthy memeber, you need to start [failure recovery](#failure-recovery).
+
+Standby-mode members will exit after etcd cluster is upgraded to v2.0:
+
+```
+Detect the cluster has been upgraded to internal API v2. Exit now.
+```
+
+When restart, standbys run in v2.0 proxy mode, which proxy user requests to etcd cluster.
+
+### Downgrade
+
+If the upgrading fails due to disk/network issues, you still can restart the upgrading process manually. However, once you upgrade etcd to internal API v2, you CANNOT downgrade it back to internal API v1. If you want to downgrade etcd in the future, please backup your v1 data dir beforehand.
+
+### Unavailable window
+
+etcd cluster is unavailable between the time that the upgrade command is sent and the time that the whole cluster is restarted.
+
+Usually, the window should be shorter than 1 minute.
+
+### In CoreOS system
+
+When you are in CoreOS system, allow-legacy mode is on automatically. Moreover, etcd process will restart automatically, so you don't need to do anything after calling upgrade command.
```

Comment:
s/you are in/running on a/
s/mode is on automatically/mode is enabled by default/
s/after calling/after running the/